# Tottenham v. Liverpool - 2024-12-22 match analysis
*This notebook intends to walk through the theory and calculations I am currently using in my model to calculate probability of 1 x 2 result and a table of over / under goal projections. I will break the calculations out into utility functions stored in `../src/utils` for ease of use in future explorations. This is to document the calculations.*

### Instructions:

1. Ensure all cells in all notebooks in `/notebooks` directory have been run.
2. Define `home_side` and `away_side` in first cell. If making a comparison with odds from `03_retrieve_odds.ipynb` also enter matchday.
3. Click 'run all'
4. Probability for 1 x 2 and over / under can be viewed in the last cell
5. *Suggestion:* copy and paste notebook to `../output/< matchday_directory >/` and rename file. IMPORTANT: Update file path in lines 14 and 15 if needed if you make a copy.



### Import League Data & Define Sides

Define 'home_side' and 'away_side'
Ensure correct output



In [35]:
import pandas as pd
from scipy.stats import poisson
import numpy as np

"""
BUG: if copying from `notebooks/03_analysis_template_notebook.ipynb` to a directory in output
for storage, ensure the file path for reading the csv will need to be updated from '../data/processed/home_table.csv'
to '../../data/processed/home_table.csv' if frames are run again.

TLDR
if you are in ./notebooks/03_analysis_tempalte.csv you may run this cell as is. If you are not, check that following
read_csv commands read from the correct directory.
"""
home_dataframe = pd.read_csv('../data/processed/home_table.csv')
away_dataframe = pd.read_csv('../data/processed/away_table.csv')

"""
!!! INPUT TEAMS & MATCHDAY HERE !!!

wrapped in quotes, eg.

home_side = "Bournemouth"
away_side = "West Ham"
matchday = "2024-12-16" # (optional)

If there is a ValueError,
(1) check spelling against the error message. Must be exact.
(2) check capitalization
(3) check that it is wrapped in quotes

!!! INPUT TEAMS & MATCHDAY HERE !!!
"""
home_side = "Tottenham"
away_side = "Liverpool"
matchday = "2024-12-22" # (optional)

# validate input
clubs = set(home_dataframe['Team'].values)
if home_side not in clubs:
    raise ValueError(f"Invalid entry for 'home_side'. Expected one of {clubs}, but received '{home_side}'")
elif away_side not in clubs:
    raise ValueError(f"Invalid entry for 'away_side'. Expected one of {clubs}, but received '{away_side}'")
else:
    print(f"Ready to calculate probability for {home_side} v. {away_side}")
    print("\nNOTEBOOK TITLE (copy and paste): \n")
    print(f"{home_side} v. {away_side} - {matchday} match analysis")

Ready to calculate probability for Tottenham v. Liverpool

NOTEBOOK TITLE (copy and paste): 

Tottenham v. Liverpool - 2024-12-22 match analysis


### Calculate Projected Goals

In [36]:
# 2. grab their gpm_scored and gpm_conceded
home_side_gpm_scored = home_dataframe.loc[
    home_dataframe["Team"] == home_side,
    "gpm_scored"
].values[0]

home_side_gpm_conceded = home_dataframe.loc[
    home_dataframe["Team"] == home_side,
    "gpm_conceded"
].values[0]

print(home_side_gpm_scored, home_side_gpm_conceded)

away_side_gpm_scored = away_dataframe.loc[
    away_dataframe["Team"] == away_side,
    "gpm_scored"
].values[0]

away_side_gpm_conceded = away_dataframe.loc[
    away_dataframe["Team"] == away_side,
    "gpm_conceded"
].values[0]

print(away_side_gpm_scored, away_side_gpm_conceded)

# 3.
epl_home_avg_gpm_scored = round(sum(home_dataframe["gpm_scored"].values) / 20, 2)
epl_home_avg_gpm_conceded = round(sum(home_dataframe["gpm_conceded"].values) / 20, 2)
print(epl_home_avg_gpm_scored, epl_home_avg_gpm_conceded)

# 4. create a ratio (attack and defense rating) of their gpm to the league average gpm
home_attack_rating = home_side_gpm_scored / epl_home_avg_gpm_scored
away_defense_rating = away_side_gpm_conceded / epl_home_avg_gpm_scored
away_attack_rating = away_side_gpm_scored / epl_home_avg_gpm_conceded
home_defense_rating = home_side_gpm_conceded / epl_home_avg_gpm_conceded

print(home_attack_rating, away_defense_rating, away_attack_rating, home_defense_rating)

# home projected goals
home_projected_goals = home_attack_rating * away_defense_rating * epl_home_avg_gpm_scored
# away projected goals
away_projected_goals = away_attack_rating * home_defense_rating * epl_home_avg_gpm_conceded
print("\n ================== projected_goals ==================== \n")
print(f"HOME {home_side} projected goals:", home_projected_goals)
print(f"AWAY {away_side} projected goals: ", away_projected_goals)

2.5 1.375
2.2857142857142856 1.1428571428571428
1.61 1.4
1.5527950310559004 0.7098491570541259 1.6326530612244898 0.9821428571428572

 ================== projected_goals ==================== 

HOME Tottenham projected goals: 1.7746228926353145
AWAY Liverpool projected goals:  2.2448979591836737


### Poisson Distribution of Goal Probabilities

In [37]:
# Poisson probability for home team goals 0 - 8
home_score_prob = []
away_score_prob = []
for i in range(9):
    prob = poisson.pmf(i, home_projected_goals)
    home_score_prob.append(prob)
for i in range(9):
    prob = poisson.pmf(i, away_projected_goals)
    away_score_prob.append(prob)

home_score_prob = np.array(home_score_prob)
away_score_prob = np.array(away_score_prob)
# print(home_score_prob)
# print(away_score_prob)

exact_score_prob = np.outer(away_score_prob, home_score_prob)
print(f"{home_side} v. {away_side}", exact_score_prob)

Tottenham v. Liverpool [[1.79615691e-02 3.18750118e-02 2.82830628e-02 1.67305902e-02
  7.42262211e-03 2.63447102e-03 7.79198765e-04 1.97540567e-04
  4.38200015e-05]
 [4.03218899e-02 7.15561489e-02 6.34925899e-02 3.75584679e-02
  1.66630292e-02 5.91411863e-03 1.74922172e-03 4.43458415e-04
  9.83714318e-05]
 [4.52592642e-02 8.03181263e-02 7.12671928e-02 4.21574639e-02
  1.87034002e-02 6.63829642e-03 1.96341213e-03 4.97759445e-04
  1.10416913e-04]
 [3.38674766e-02 6.01019993e-02 5.33291919e-02 3.15464016e-02
  1.39957416e-02 4.96743269e-03 1.46921996e-03 3.72473054e-04
  8.26249011e-05]
 [1.90072573e-02 3.37307139e-02 2.99296485e-02 1.77046131e-02
  7.85475294e-03 2.78784488e-03 8.24562224e-04 2.09041000e-04
  4.63711180e-05]
 [8.53387061e-03 1.51444021e-02 1.34378014e-02 7.94900998e-03
  3.52662377e-03 1.25168546e-03 3.70211611e-04 9.38551428e-05
  2.08196856e-05]
 [3.19294479e-03 5.66627291e-03 5.02774881e-03 2.97411938e-03
  1.31948508e-03 4.68317687e-04 1.38514548e-04 3.51158697e-05
 

### Under Market

In [38]:
thresholds = [0.5, 1.5, 2.5, 3.5]
sums = []

for t in thresholds:
    mask = np.add.outer(
        np.arange(
            exact_score_prob.shape[0]
        ),
        np.arange(
            exact_score_prob.shape[1]
        )
    ) <= t
    # print(mask)
    sums.append(exact_score_prob[mask].sum())

# print(sums)
under_market_df = pd.DataFrame({
    "Goals": thresholds,
    "Prob": sums
})
under_market_df["Implied Odds"] = 1 / under_market_df["Prob"]
print("~  ~ under market ~  ~\n", f"{home_side} v. {away_side}\n", under_market_df)

~  ~ under market ~  ~
 Tottenham v. Liverpool
    Goals      Prob  Implied Odds
0    0.5  0.017962     55.674423
1    1.5  0.090158     11.091581
2    2.5  0.235257      4.250672
3    3.5  0.429666      2.327391


In [39]:
thresholds = [0.5, 1.5, 2.5, 3.5]
sums = []

for t in thresholds:
    mask = np.add.outer(
        np.arange(
            exact_score_prob.shape[0]
        ),
        np.arange(
            exact_score_prob.shape[1]
        )
    ) >= t
    sums.append(exact_score_prob[mask].sum())

# print(sums)
over_market_df = pd.DataFrame({
    "Goals": thresholds,
    "Prob": sums
})

over_market_df["Implied Odds"] = 1 / over_market_df["Prob"]
print("~  ~ over market ~  ~\n", f"{home_side} v. {away_side}\n", over_market_df)

~  ~ over market ~  ~
 Tottenham v. Liverpool
    Goals      Prob  Implied Odds
0    0.5  0.981398      1.018954
1    1.5  0.909201      1.099866
2    2.5  0.764103      1.308724
3    3.5  0.569694      1.755328


### 1 x 2 Probability

In [40]:
# Poisson probability for home team goals 0 - 8
home_score_prob = []
away_score_prob = []
for i in range(9):
    prob = poisson.pmf(i, home_projected_goals)
    home_score_prob.append(prob)
for i in range(9):
    prob = poisson.pmf(i, away_projected_goals)
    away_score_prob.append(prob)

home_score_prob = np.array(home_score_prob)
away_score_prob = np.array(away_score_prob)
# print(home_score_prob)
# print(away_score_prob)

exact_score_prob = np.outer(away_score_prob, home_score_prob)

# print(exact_score_prob)
home_win_prob = np.sum(np.triu(exact_score_prob, k=1))
away_win_prob = np.sum(np.tril(exact_score_prob, k=-1))
draw_prob = np.trace(exact_score_prob)

# prepare data in readable dataframe
data = {
    "winner": [f"{home_side} (1)", "Draw (x)", f"{away_side} (2)"],
    "prob": [home_win_prob, draw_prob, away_win_prob],
}
result_df = pd.DataFrame(data)
result_df['implied_odds'] = 1 / result_df['prob']
print("      ~  ~ moneyline market ~  ~\n", result_df)

      ~  ~ moneyline market ~  ~
           winner      prob  implied_odds
0  Tottenham (1)  0.309242      3.233714
1       Draw (x)  0.201588      4.960607
2  Liverpool (2)  0.488530      2.046958


# Full Results Summary:

In [41]:
# helper function to translate discrepency between Understat and Odds-Api naming
def translate(string):

    translations = {
        "Brighton": "Brighton and Hove Albion",
        "Ipswich": "Ipswich Town",
        "Leicester": "Leicester City",
        "Tottenham": "Tottenham Hotspur",
        "West Ham": "West Ham United"
    }

    if string in translations:
        return translations[string]
    return string

home_side, away_side = translate(home_side), translate(away_side)

match_string = f"{home_side} v. {away_side}"

print(f"CALCULATED PROBABILITIES FOR {home_side} v. {away_side}")
print("\n\n        ~ ~ moneyline market ~ ~\n\n", result_df)
print("\n\n          ~ ~ over market ~ ~ \n\n", over_market_df)
print("\n\n          ~ ~ under market ~ ~ \n\n", under_market_df)

try:
    odds_df = pd.read_csv(f"../data/processed/odds/odds_matchday_{matchday}.csv")
    filtered_odds = odds_df.loc[odds_df["match"] == match_string]
    implied = {
        "1": 1 / filtered_odds["1"],
        "x": 1 / filtered_odds["x"],
        "2": 1 / filtered_odds["2"],
        "point": filtered_odds["ovr_und_point"],
        "over": 1 / filtered_odds["over"],
        "under": 1 / filtered_odds["under"]
    }
    implied_odds_df = pd.DataFrame(implied)

    print(f"\n ~~ MyBookie Odds for {match_string} ~~\n",filtered_odds.drop(columns=["match"]))
    print(f"\n ~~ MyBookie Odds IMPLIED for {match_string} ~~\n", implied_odds_df)
except:
    print(f"No data for sportsbook odds on matchday {matchday}. To view against real odds, run `notebooks/03_retrieve_odds.ipynb` ran correctly and ensure matchday is correct in the first cell of this notebook")


CALCULATED PROBABILITIES FOR Tottenham Hotspur v. Liverpool


        ~ ~ moneyline market ~ ~

           winner      prob  implied_odds
0  Tottenham (1)  0.309242      3.233714
1       Draw (x)  0.201588      4.960607
2  Liverpool (2)  0.488530      2.046958


          ~ ~ over market ~ ~ 

    Goals      Prob  Implied Odds
0    0.5  0.981398      1.018954
1    1.5  0.909201      1.099866
2    2.5  0.764103      1.308724
3    3.5  0.569694      1.755328


          ~ ~ under market ~ ~ 

    Goals      Prob  Implied Odds
0    0.5  0.017962     55.674423
1    1.5  0.090158     11.091581
2    2.5  0.235257      4.250672
3    3.5  0.429666      2.327391

 ~~ MyBookie Odds for Tottenham Hotspur v. Liverpool ~~
       1    x     2  ovr_und_point  over  under
4  3.95  4.2  1.72            3.5  1.71   2.06

 ~~ MyBookie Odds IMPLIED for Tottenham Hotspur v. Liverpool ~~
           1         x         2  point      over     under
4  0.253165  0.238095  0.581395    3.5  0.584795  0.485437


# Conclusion and Pick
Possibly an over 3.5 is in the cards. I want to show some stones and back the coys, it's been all negative so far this weekend. We'll see how the first picks go and make a call.
### My Pick:
Tentative: **O 3.5 @1.71** -- *1 unit*
